### ISSUE TITLE
golang should provide an API to call `go list`

### ISSUE THREAD

**Issue by nim-nim** [golang/go#30505]
```console
$ go version
go version go1.12 linux/amd64
```
Calling `exec.Command("go", "list", "-json"…` just to re-parse the produced json file in a Go structure is not terribly elegant. Moreover, the `go list` cli does not expose all the options necessary to control the desired output (for example, the build context to operate on, that may be different from the one infered from the local system).

Go should provide an API to perform `go list` operations from Go code without calling a CLI command.

**Comment by mvdan** [comment]
You very likely want https://godoc.org/golang.org/x/tools/go/packages.

**Comment by ianlancetaylor** [comment]
Calling `go list` (which go/packages does under the hood) may not be elegant, but it will work across Go versions.  The layout of the information used by `go list` can change across Go releases.  That means that if you build a Go program with release 1.N, and then later run it on a system that is using release 1.M, you can get different results.  It's true that we could put in the effort to maintain backward compatibility (the case where M < N), but we could not maintain forward compatibility (M > N) without either a lot of work or tying our hands severely.  We've decided that the simplest and most reliable approach is to require invoking `go list`, and we will make sure that that works.

In short, we are going to use the solution that works even if it does not seem elegant.

**Comment by nim-nim** [comment]
A solution that works would be awfully nice :)

More seriously, go/packages is package-oriented, while go is moving to modules (set of packages), so go/packages is not sufficient anymore, and calling go list by hand is quite tricky to do right.

**Comment by ianlancetaylor** [comment]
If there are problems with using go/packages with modules, then please file bugs against go/packages.

**Comment by mvdan** [comment]
I guess @nim-nim means an API to use commands like `go list -m`. `go/packages` is to load packages, not query modules, so it's just a different purpose.

@nim-nim I presume what you want would be a Go wrapper around `go list -m -json`. I personally don't think that's critical enough to warrant being a package like `x/tools/go/modules`, but perhaps you can file a proposal and try to convince the tooling team.

**Comment by nim-nim** [comment]
@mvdan `go list -json ./...` is nice because it walks all the subpackages in a specific filesystem tree, and gives you both package info (package file lists) and module info (the go.mod the package depends on and the module name)

From my look at `go/packages`, it does not provide this level of synthesis right now

**Comment by mvdan** [comment]
Because `go/packages` is about packages, not modules. It will support Bazel and Blaze too, for example, which build Go packages but do not use modules at all.

The moment you want to query modules, you're working with `go list`, which is lower-level than what `go/packages` does.
